### PR TITLE
CA-290696: Update task to un-cancellable after xenopsd notify xapi

### PR DESCRIPTION
**Pull request by krizex**
Set the task to un-cancellable after xenopsd updated it to un-cancellable

Signed-off-by: Yang Qian <yang.qian@citrix.com>

### CODE REVIEW

**Comment by krizex** [comment]
This PR depends on https://github.com/xapi-project/xcp-idl/pull/240 and https://github.com/xapi-project/xenopsd/pull/550

**Comment by lindig** [comment]
Please update the commit message such that the topic does not overflow:
```
CA-290696 - block VM migration cancelation after VM suspended
```

The logic looks sounds to me.

File: ocaml/xapi/xapi_xenops.ml [hunk 1]
```diff
@@ -2058,7 +2058,15 @@ let update_task ~__context queue_name id =
...
     match task_t.Task.state with
     | Task.Pending x ->
-      Db.Task.set_progress ~__context ~self ~value:x
+      Db.Task.set_progress ~__context ~self ~value:x;
+      if not task_t.Task.cancellable then begin
+        let allowed_operations = Db.Task.get_allowed_operations ~__context ~self in
+        if List.mem `cancel allowed_operations then begin
+          let allowed_operations' = List.filter (fun x -> x != `cancel) allowed_operations in
+          debug "Set task %s to not cancellable." (Ref.string_of self);
+          Db.Task.set_allowed_operations ~__context ~self ~value:allowed_operations'
```
Comment (robhoes) [inline ocaml/xapi/xapi_xenops.ml]: What about `TaskHelper.set_not_cancellable ~__context`?

Comment (krizex) [reply ocaml/xapi/xapi_xenops.ml]: `TaskHelper.set_not_cancellable ~__context` will set the task of the `__context` to not cancellable, but in this case, the task we have to update status not in `__context`

File: ocaml/xapi/xapi_xenops.ml [hunk 2]
```diff
@@ -2058,7 +2058,15 @@ let update_task ~__context queue_name id =
     let task_t = Client.TASK.stat dbg id in
     match task_t.Task.state with
     | Task.Pending x ->
-      Db.Task.set_progress ~__context ~self ~value:x
+      Db.Task.set_progress ~__context ~self ~value:x;
+      if not task_t.Task.cancellable then begin
+        let allowed_operations = Db.Task.get_allowed_operations ~__context ~self in
+        if List.mem `cancel allowed_operations then begin
+          let allowed_operations' = List.filter (fun x -> x != `cancel) allowed_operations in
```
Comment (robhoes) [inline ocaml/xapi/xapi_xenops.ml]: You should use `<>` rather than `!=`.

File: ocaml/xapi/xapi_xenops.ml [hunk 3]
```diff
@@ -2058,7 +2058,15 @@ let update_task ~__context queue_name id =
...
+          debug "Set task %s to not cancellable." (Ref.string_of self);
```
Comment (robhoes) [inline ocaml/xapi/xapi_xenops.ml]: Here you can use `Ref.really_pretty_and_small` to get a task ref as in the logs (e.g. `R:xxxxxxxx`), which makes it easier to follow.

File: ocaml/xapi/xapi_xenops.ml (hunk 2 above)
Comment (krizex) [reply ocaml/xapi/xapi_xenops.ml]: Oh, my bad.

File: ocaml/xapi/xapi_xenops.ml (hunk 3 above)
Comment (krizex) [reply ocaml/xapi/xapi_xenops.ml]: Does this function only suitable for task ref?

Comment (robhoes) [reply ocaml/xapi/xapi_xenops.ml]: It works on any ref, but we seem to use it just for tasks. Also, the prefixes `R` (a "real" ref for something in the DB) and `D` (a dummy ref; nothing in the DB) are only relevant for tasks.

**Comment by robhoes** [comment]
This is good, but can't be merged until the xcp-idl patch has been rebased and merged.

**Pull request merged**